### DISCUSSION TITLE
Regular expressions with textedit?

### DISCUSSION THREAD

**Discussion by saward** [emilk/egui#1200] [Q&A] [answered]
Hi!  I'm wondering what would be a good option for applying a regular expression/pattern match to a textedit, so that users can't enter text other than that allowed?

I think I could do this by checking the response for changed(), but in that case there's a slight flicker because presumably there's a frame where the text is changed, before I revert it to the last permitted state.  I can live with the flicker, but interested to know if there's a better approach.

**Comment by emilk** [comment, marked as answer]
There are two situations to consider:

* Allowed character (e.g. only entering numbers)
* Allowed formats (e.g. only allowing valid email addresses)

For the first case, I think we should add a filter to what characters `TextEdit` accepts, i.e. just a `Fn(char) -> bool` passed to one of the `TextEdit` builder functions. If the users tried to enter characters outside this filter, they won't be accepted.

For the second case, I think it is better to leave the validation to the egui user. How to respond to invalid input depends highly on the situation - sometimes you want to color the `TextEdit` red, or show an error text next to it.

**Comment by saward** [comment]
Hi, thanks for answering!  My situation is a bit more complex, like the email case (in my case, ensuring, e.g., that money entered has no more than one decimal).  I could do the validation and provide an error rather than prohibit invalid input.  I think that's a great approach that would work with egui in its current form.

Thinking about changes to egui, I had been thinking of a way to pass a regular expression to `TextEdit`, and if the newly entered character would not match the regex then it would reject that character.  Similar to the first solution you propose, except passing a regex instead of character filters.  Do you think that would be a useful solution?

**Reply by emilk** [reply]
A character filter of the form `Fn(char) -> bool` is more general than a regex, but more importantly it does not require egui to depend on the enormous `regex` crate.

**Comment by saward** [comment]
Ah, maybe I misunderstood the filter option.  So the user would provide a function of the form `Fn(char) -> bool`, and then it would be used to decide whether to reject a character or not?  And therefore, the user (such as myself) could use a regex inside that provided function?

**Reply by emilk** [reply]
yup

**Reply by saward** [reply]
Sounds like it would do the trick then!

**Reply by Mingun** [reply]
Probably it is better to use a [`Pattern`](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html) trait for this, but right now it is nightly-only
